### PR TITLE
Properly manage `fetcher.done` with `dataloader_iter`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Fixes calculation of `data_fetcher.done` to be consistent with regular data fetchers.
- Makes `dataloader_iter.{done,length,fetched}` public

cc @justusschock @awaelchli @borda

### CODE REVIEW

File: tests/tests_pytorch/loops/test_fetchers.py [hunk 1]
```diff
@@ -294,7 +301,6 @@ def training_step(self, dataloader_iter: Iterator) -> STEP_OUTPUT:
         self.optimizers().zero_grad()
 
         self.batch_i_handle = batch_ip1_handle
-        self.num_batches_processed += 1
```
Comment (carmocca) [inline tests/tests_pytorch/loops/test_fetchers.py:297]: This was wrong because it doesn't take into account that the first `training_step` calls `next()` twice

File: tests/tests_pytorch/loops/test_fetchers.py [hunk 2]
```diff
@@ -115,11 +115,14 @@ def __len__(self):
...
 @pytest.mark.parametrize("prefetch_batches", list(range(2)))
 def test_empty_prefetch_iterator(dataset_cls, prefetch_batches):
-    loader = DataLoader(dataset_cls())
+    loader = CombinedLoader(DataLoader(dataset_cls()))
     fetcher = _PrefetchDataFetcher(prefetch_batches=prefetch_batches)
     fetcher.setup(loader)
 
-    assert not fetcher.done
+    if dataset_cls is EmptySizedDataset:
+        assert fetcher.done  # for 0 length sized datasets we know we're done already
```
Comment (carmocca) [inline tests/tests_pytorch/loops/test_fetchers.py:123]: this is an improvement for the edge case of 0 length sized datasets

File: tests/tests_pytorch/loops/test_fetchers.py [hunk 3]
```diff
@@ -43,17 +43,17 @@ def __getitem__(self, idx):
...
+def test_prefetch_iterator(multiple_iterables, dataset_cls, prefetch_batches):
     fetcher = _PrefetchDataFetcher(prefetch_batches=prefetch_batches)
     assert fetcher.prefetch_batches == prefetch_batches
 
-    if use_combined_loader:
+    if multiple_iterables:
         loader = CombinedLoader([DataLoader(dataset_cls()), DataLoader(dataset_cls())])
     else:
-        loader = DataLoader(dataset_cls())
+        loader = CombinedLoader(DataLoader(dataset_cls()))
```
Comment (carmocca) [inline tests/tests_pytorch/loops/test_fetchers.py:56]: Minor test fix. `fetcher.setup` always expects a `CombinedLoader`

File: src/lightning/pytorch/loops/fetchers.py [hunk 4]
```diff
@@ -125,27 +129,16 @@ def __next__(self) -> Any:
                 self.done = not self.batches
         elif not self.done:
             # this will run only when no pre-fetching was done.
-            try:
-                self._fetch_next_batch(self.iterator)
-                # consume the batch we just fetched
-                batch = self.batches.pop(0)
-            except StopIteration as ex:
```
Comment (carmocca) [inline src/lightning/pytorch/loops/fetchers.py:132]: This `StopIteration` is now managed in `self._fetch_next_batch`

File: src/lightning/pytorch/loops/fetchers.py [hunk 5]
```diff
@@ -107,7 +112,6 @@ def __iter__(self) -> "_PrefetchDataFetcher":
             except StopIteration:
                 # this would only happen when prefetch_batches > the number of batches available and makes
                 # `__next__` jump directly to the empty iterator case without trying to fetch again
-                self.done = True
```
Comment (carmocca) [inline src/lightning/pytorch/loops/fetchers.py:110]: This is now managed in `self._fetch_next_batch`

File: src/lightning/pytorch/loops/fetchers.py [hunk 6]
```diff
@@ -87,15 +97,10 @@ def __init__(self, prefetch_batches: int = 1) -> None:
...
-        self._len: Optional[int] = None
-
-    def setup(self, combined_loader: CombinedLoader) -> None:
-        super().setup(combined_loader)
-        self._len = sized_len(combined_loader)
 
     def __iter__(self) -> "_PrefetchDataFetcher":
         super().__iter__()
-        if self._len is not None:
+        if self.length is not None:
```
Comment (carmocca) [inline src/lightning/pytorch/loops/fetchers.py:103]: API promotion

File: src/lightning/pytorch/loops/fetchers.py [hunk 7]
```diff
@@ -87,15 +97,10 @@ def __init__(self, prefetch_batches: int = 1) -> None:
             raise ValueError("`prefetch_batches` should at least be 0.")
         self.prefetch_batches = prefetch_batches
         self.batches: List[Any] = []
-        self._len: Optional[int] = None
```
Comment (carmocca) [inline src/lightning/pytorch/loops/fetchers.py:90]: This is now in the parent

File: tests/tests_pytorch/loops/test_fetchers.py [hunk 8]
```diff
@@ -498,3 +506,40 @@ def training_step(self, dataloader_iter):
...
+        next(fetcher)
+    assert fetcher.done
+
+    loader = CombinedLoader(iterable)
+    fetcher = _DataLoaderIterDataFetcher()
+    fetcher.setup(loader)
+    iter(fetcher)
+    assert not fetcher.done
+    for i in range(5):  # doesn't matter how many times you next this, the iter itself needs to be consumed
+        dataloader_iter = next(fetcher)
```
Comment (awaelchli) [inline tests/tests_pytorch/loops/test_fetchers.py:534]: maybe one could assert also 

```py
assert next(fetcher) is next(fetcher)
```

File: src/lightning/pytorch/loops/fetchers.py [hunk 9]
```diff
@@ -42,24 +43,28 @@ def combined_loader(self) -> CombinedLoader:
 
     def setup(self, combined_loader: CombinedLoader) -> None:
         self._combined_loader = combined_loader
+        self.length = sized_len(combined_loader)
+        self.done = self.length == 0
```
Comment (awaelchli) [inline src/lightning/pytorch/loops/fetchers.py:47]: should we add a test case for this?

File: tests/tests_pytorch/loops/test_fetchers.py [hunk 10]
```diff
@@ -192,12 +195,14 @@ def training_step(self, dataloader_iter, batch_idx):
                 opt.step()
 
         def on_train_epoch_end(self):
-            assert self.trainer.fit_loop.epoch_loop.batch_progress.current.ready == 33
+            # since the dataset is sized, the loop stops at the limit even though the training_step controls the
+            # consumption of batches
```
Comment (awaelchli) [inline tests/tests_pytorch/loops/test_fetchers.py:199]: Does that mean we no longer enter the *_step() method where the StopIteration would normally occur? (for the sized iterator case)

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/loops/fetchers.py (hunk 9 above)
Comment (carmocca) [reply src/lightning/pytorch/loops/fetchers.py:47]: https://github.com/Lightning-AI/lightning/pull/18376/files#r1303403909 covers it

File: tests/tests_pytorch/loops/test_fetchers.py (hunk 10 above)
Comment (carmocca) [reply tests/tests_pytorch/loops/test_fetchers.py:199]: That's right. This also means https://docs.google.com/document/d/1DQxvUdkxdh3jmWorA5g12LeOOU0tcJ7d0uqN5LyvISQ/edit?disco=AAAA4XIqT6M

File: src/lightning/pytorch/loops/fetchers.py (hunk 4 above)
Comment (Borda) [reply src/lightning/pytorch/loops/fetchers.py:132]: lets add it as comment...

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/loops/fetchers.py [hunk 11]
```diff
@@ -42,24 +43,28 @@ def combined_loader(self) -> CombinedLoader:
 
     def setup(self, combined_loader: CombinedLoader) -> None:
         self._combined_loader = combined_loader
+        self.length = sized_len(combined_loader)
```
Comment (awaelchli) [inline src/lightning/pytorch/loops/fetchers.py:46]: I'm exploring some ideas how we could make this length respect the `limit_x_batches` settings from the loop: #18436
